peer/peerlist/v2/list.go

```diff
@@ -491,22 +491,30 @@ func (pl *List) getThunk(pid peer.Identifier) *peerThunk {
 	return pl.unavailablePeers[pid.Identifier()]
 }
 
-// notifyStatusChanged gets called by peer thunks
+// notifyStatusChanged gets called by peer thunks.
+// Change notifications must be sent outside the list lock since handlers may
+// call locking list methods.
 func (pl *List) notifyStatusChanged(pid peer.Identifier) {
 	pl.lock.Lock()
-	defer pl.lock.Unlock()
 
 	if t := pl.availablePeers[pid.Identifier()]; t != nil {
```

Comment:
We should avoid having so many different paths between `Lock` and `Unlock`, we have a couple of other options:

```go
pl.lock.Lock()
isAvailable := pl.availablePeers[pid.Identifier()]
isUnavailable := pl.unavailablePeers[pid.Identifier()]
pl.lock.Unlock()

if isAvailable {...}
if isUnavailable { ... }
```

Or if we don't want to do 2 map lookups, we just move the `isUnavailable` check inside
```go
pl.lock.Lock()
isAvailable := pl.availablePeers[pid.Identifier()]
var isUnavailable bool
if !isAvailable {
  isUnavailable= pl.unavailablePeers[pid.Identifier()]
}
pl.lock.Unlock()
```